mlir/unittests/IR/InterfaceTest.cpp

```diff
@@ -84,3 +87,40 @@ TEST(InterfaceTest, TestImplicitConversion) {
   typeA = typeB;
   EXPECT_EQ(typeA, typeB);
 }
+
+TEST(OperationInterfaceTest, CastOpToInterface) {
+  DialectRegistry registry;
+  MLIRContext ctx;
+
+  const char *ir = R"MLIR(
+      func.func @map(%arg : tensor<1xi64>) {
+        %0 = arith.constant dense<[10]> : tensor<1xi64>
+        %1 = arith.addi %arg, %0 : tensor<1xi64>
+        return
+      }
+    )MLIR";
+
+  registry.insert<func::FuncDialect, arith::ArithDialect>();
+  ctx.appendDialectRegistry(registry);
+  OwningOpRef<ModuleOp> module = parseSourceString<ModuleOp>(ir, &ctx);
+  Operation &op = cast<func::FuncOp>(module->front()).getBody().front().front();
+
+  OpAsmOpInterface interface = llvm::cast<OpAsmOpInterface>(op);
+
+  bool constantOp =
+      llvm::TypeSwitch<OpAsmOpInterface, bool>(interface)
+          .Case<VectorUnrollOpInterface, arith::ConstantOp>([&](auto op) {
+            return std::is_same_v<decltype(op), arith::ConstantOp>;
+          });
+
+  EXPECT_TRUE(constantOp);
+
+  EXPECT_FALSE(llvm::isa<VectorUnrollOpInterface>(interface));
+  EXPECT_FALSE(llvm::dyn_cast<VectorUnrollOpInterface>(interface));
+
+  EXPECT_TRUE(llvm::isa<InferTypeOpInterface>(interface));
+  EXPECT_TRUE(llvm::dyn_cast<InferTypeOpInterface>(interface));
+
+  EXPECT_TRUE(llvm::isa<OpAsmOpInterface>(interface));
+  EXPECT_TRUE(llvm::dyn_cast<OpAsmOpInterface>(interface));
+}
```

Comment:
Would this test really fail before the fix? 
Or would it trigger UBSAN in a way that no other test would have before because we don't cast from an interface in general?

Reply:
If there is no such change, the test case will fail to compile on my local machine. (gcc (Ubuntu 9.4.0-1ubuntu1~20.04.3) 9.4.0). @ingomueller-net Can you help me check if there are any differences between the tests here and the ones you provided in the [issue](https://github.com/llvm/llvm-project/issues/86647)?

Reply:
On first sight, this looks good! If I remember/reconstructed things correctly, the main breakage resulted from having an interface argument instead of `Operation*`.

I have just started to CI jobs in the hope to confirm that your patch fixes my original problem:

1. [This](https://github.com/iree-org/iree-llvm-sandbox/actions/runs/9857834962/job/27217780796?pr=850) job is for a commit that reverts my work-around by changing the typeswitch argument type back to the interface type and should thus *fail*.
2. [This](https://github.com/iree-org/iree-llvm-sandbox/actions/runs/9857967690/job/27218209768?pr=850) job has an additional commit that applies the diff from this PR as a patch on the MLIR source code and should thus *succeed* in the hope that your patch fixes the previously introduced breakage.

I haven't worked on that project for several weeks so the CI caches are flushed and CI will likely take 1-2 hours. I'll report back later when I'll be able to check the results but in case I don't manage to do that today, I am sharing the links now.

Reply:
Thanks very much.If I use `&op` typeyof  `Operation*` instread of `interface` typeyof `OpAsmOpInterface` in my test case, compiler will report error:`could not convert Operation* to OpAsmOpInterface`.I think there may be some practices in IREE that cause implicit conversions to occur.

Reply:
The CI jobs finished and failed both. I have just rerun them now with the tracing from #86647. Unfortunately, the version with the patch ([CI job](https://github.com/iree-org/iree-llvm-sandbox/actions/runs/9870555351/job/27256554874?pr=850)) still fails and the tracing shows that the type switch takes the wrong branch:

```
auto (anonymous namespace)::exportOperation(mlir::substrait::ExpressionOpInterface)::(anonymous class)::operator()(auto) const [op:auto = mlir::substrait::FieldReferenceOp]
%3 = substrait.literal -1 : si1
```

I also made sure that `OpDefinition.h` is included by including it directly but CI still [fails](https://github.com/iree-org/iree-llvm-sandbox/actions/runs/9870726034/job/27257087019?pr=850) with the same tracing output.

Reply:
Yes.When I use below test cast, it will due to cast failed then dereference nullptr and core dump, I think I need to implement the `CastInfo<Op<ConcreteType, Traints...>, Interface>`
```
  bool constantOp =
      llvm::TypeSwitch<OpAsmOpInterface, bool>(interface)
          .Case<arith::AddIOp, arith::ConstantOp>([&](auto op) {
            return std::is_same_v<decltype(op), arith::ConstantOp>;
          });
```

Reply:
I have fixed and eliminated compiler warnings `returning reference to local temporary object`, as well as fixed the issue of `OpInterface` crashing when converting to `Op`. There seems to be an environmental issue with i`ree-llvm-sandbox` on my local machine, and I am planning to submit this fix for this project to try. Currently, I need to refine this fix again.

Reply:
@ingomueller-net Could you help to try this [PR](https://github.com/iree-org/iree-llvm-sandbox/pull/851)?